tests/backup/backup_kubevirt_test.go

```diff
@@ -2150,3 +2150,486 @@ var _ = Describe("{KubevirtVMRestoreWithAfterChangingVMConfig}", func() {
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creation of cloud credential named [%s] for org [%s] with [%s] as provider", cloudCredName, BackupOrgID, provider))
+				err = CreateBackupLocation(provider, backupLocationName, backupLocationUID, cloudCredName, cloudCredUID, getGlobalBucketName(provider), BackupOrgID, "", true)
+				dash.VerifyFatal(err, nil, "Creation of backup location")
+			}
+		})
+
+		Step("Register clusters for backup & restore", func() {
+			log.InfoD("Registering clusters for backup & restore")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Unable to fetch px-central-admin ctx")
+
+			err = CreateApplicationClusters(BackupOrgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination clusters")
+
+			clusterStatus, err := Inst().Backup.GetClusterStatus(BackupOrgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+
+			sourceClusterUID, err = Inst().Backup.GetClusterUID(ctx, BackupOrgID, SourceClusterName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+
+			clusterStatus, err = Inst().Backup.GetClusterStatus(BackupOrgID, DestinationClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", DestinationClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", DestinationClusterName))
+		})
+
+		Step("Create schedule policies", func() {
+			log.InfoD("Creating schedule policies")
+			periodicSchedulePolicyInfo := Inst().Backup.CreateIntervalSchedulePolicy(5, 15, 2)
+			periodicPolicyStatus := Inst().Backup.BackupSchedulePolicy(periodicPolicyName, uuid.New(), BackupOrgID, periodicSchedulePolicyInfo)
+			dash.VerifyFatal(periodicPolicyStatus, nil, fmt.Sprintf("Creation of periodic schedule policy - %s", periodicPolicyName))
+		})
+
+		Step("Create pre & post exec rules", func() {
+			log.InfoD("Creating pre & post exec rules")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			// Application pre & post exec rules
+			preRuleName, postRuleName, err = CreateRuleForBackupWithMultipleApplications(BackupOrgID, Inst().AppList, ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Creation of pre & post exec rules"))
+			if preRuleName != "" {
+				preRuleNames = append(preRuleNames, preRuleName)
+			}
+			if postRuleName != "" {
+				postRuleNames = append(postRuleNames, postRuleName)
+			}
+			// Vm freeze and unfreeze rules
+			for _, appCtx := range scheduledAppContexts {
+				vms, err := GetAllVMsInNamespace(appCtx.ScheduleOptions.Namespace)
+				log.FailOnError(err, "Failed to get VMs in namespace - %s", appCtx.ScheduleOptions.Namespace)
+				allVMs = append(allVMs, vms...)
+			}
```

Comment:
Because multiple apps exist within a single namespace, duplicates might appear in allVMs as we iterate through all contexts. However, we have hardcoded the app list in this test case where one app has no VMs, and the test case will likely pass. Ideally, we should implement a filter for namespace.

How about?

In backup_helper.go

```go
// GetUniqueNamespacesFromScheduledContexts returns a slice of unique namespaces from scheduled contexts
func GetUniqueNamespacesFromScheduledContexts(scheduledAppContexts []*scheduler.Context) []string {
    namespaceSet := make(map[string]struct{})
    var uniqueNamespaces []string
    for _, ctx := range scheduledAppContexts {
        namespace := ctx.ScheduleOptions.Namespace
        if _, exists := namespaceSet[namespace]; !exists {
            namespaceSet[namespace] = struct{}{}
            uniqueNamespaces = append(uniqueNamespaces, namespace)
        }
    }
    return uniqueNamespaces
}
```
In test case

```go
uniqueSourceClusterNamespaces := GetUniqueNamespacesFromScheduledContexts(scheduledAppContexts)
for _, namespace := range uniqueSourceClusterNamespaces {
	vms, err := GetAllVMsInNamespace(namespace)
	log.FailOnError(err, "Failed to get VMs in namespace - %s", namespace)
	allVMs = append(allVMs, vms...)
}
```

Cc: @mkoppal-px

Reply:
@kphalgun-px addressed by looping the namespaces.

Reply:
Thank you.